Maps/ArchipelagoCampaign/WoL/ap_liberation_day.SC2Map/MapScript.galaxy

```diff
@@ -670,25 +682,151 @@ void gt_SixthBonusObjectiveDestroyed_Init () {
-    TriggerAddEventUnitDied(gt_TransportDestroyed, UnitRefFromUnit(UnitFromId(331)));
+void gt_ObjectiveCallTheSpecialDeliveryComplete_Init () {
+    gt_ObjectiveCallTheSpecialDeliveryComplete = TriggerCreate("gt_ObjectiveCallTheSpecialDeliveryComplete_Func");
+}
+
+//--------------------------------------------------------------------------------------------------
+// Trigger: Objective Transport Create
+//--------------------------------------------------------------------------------------------------
+bool gt_ObjectiveTransportCreate_Func (bool testConds, bool runActions) {
+    // Variable Declarations
+    int lv_category;
+
+    // Automatic Variable Declarations
+    // Variable Initialization
+    lv_category = lib5BD4895D_ge_APObjectiveCategory_Extra;
+
+    // Conditions
+    if (testConds) {
+        if (!((TriggerIsEnabled(TriggerGetCurrent()) == true))) {
+            return false;
+        }
+
+        if (!((lib5BD4895D_gf_AP_Core_isLocationCollectable(gv_LOCATION_ID_TRANSPORT) == true))) {
+            return false;
+        }
+    }
+
+    // Actions
+    if (!runActions) {
+        return true;
+    }
+
+    TriggerEnable(TriggerGetCurrent(), false);
+    ObjectiveCreate(StringExternal("Param/Value/EB9D7884"), StringToText(""), c_objectiveStateActive, false);
+    lib5BD4895D_gf_AP_Core_addCategoryObjectiveSimple(ObjectiveLastCreated(), lv_category, 1);
+    gv_objectiveTransport = ObjectiveLastCreated();
+    Wait(libCamp_gv_campaignObjectiveDelay, c_timeGame);
+    return true;
+}
+
+//--------------------------------------------------------------------------------------------------
+void gt_ObjectiveTransportCreate_Init () {
+    gt_ObjectiveTransportCreate = TriggerCreate("gt_ObjectiveTransportCreate_Func");
+    TriggerAddEventUnitRegion(gt_ObjectiveTransportCreate, null, RegionFromId(21), true);
+}
+
+//--------------------------------------------------------------------------------------------------
+// Trigger: Objective Transport Complete
+//--------------------------------------------------------------------------------------------------
+bool gt_ObjectiveTransportComplete_Func (bool testConds, bool runActions) {
+    // Automatic Variable Declarations
+    // Conditions
+    if (testConds) {
+        if (!((TriggerIsEnabled(TriggerGetCurrent()) == true))) {
+            return false;
+        }
+
+        if (!((lib5BD4895D_gf_AP_Core_isLocationCollectable(gv_LOCATION_ID_TRANSPORT) == true))) {
+            return false;
+        }
+    }
+
+    // Actions
+    if (!runActions) {
+        return true;
+    }
+
+    TriggerEnable(TriggerGetCurrent(), false);
+    TriggerExecute(gt_ObjectiveTransportCreate, true, false);
+    lib5BD4895D_gf_AP_Core_setObjectiveChecks(gv_objectiveTransport, 0);
+    lib5BD4895D_gf_AP_Core_setObjectiveCompleted(gv_objectiveTransport);
+    lib5BD4895D_gf_AP_Core_sendLocationLocal(gv_LOCATION_ID_TRANSPORT);
+    return true;

```

Comment:
On reading this again, I wonder if these sorts of checks (`isLocationCollectable(gv_LOCATION_ID_TRANSPORT) == true`) are better handled on the mod-side. ie, the maps simply specify and objective, what locations it's linked to, and whether it should display even if it has no checks, then the mod can handle all the details on whether to display it / how many checks to display.

I just feel like you're going to have to add over 600 of these if checks, and if we continue down this route there still won't be any connection in AP core mod between location index and objective struct.

Reply:
I'm imagining something like this:

## State / configuration
AP core maintains a list of objectives (configured by the map), a list of locations that can be collected (supplied by the client on startup). Objectives contain data for:
* What locations they count towards
* Whether they show if there are no locations (ie in-mission progression vs things that are AP-location-only)
* Normal stuff like counter/timer/stuff we already configure

## Internal API
* `AP_Core_locationToObjective(int locationId) -> objective`
* `AP_Core_getObjectiveLocations(objective objective) -> int[]`
* `AP_Core_updateAllObjectiveChecksCounters() -> void`
* `AP_Core_signalLocationId(int locationId) -> void`
  * Sends the location received signal to the client if the location is collectable
  * Updates the `[X checks remain]` counter on the associated objective if necessary
  * I'm not sure if this should handle counters as well or if counters should be handled with a separate function called right next to this function in the map script if necessary

## Usage in map file
The objective categorization largely doesn't change, except some extra information must be passed to objective constructors:
* `int[]` / `int` (depending on constructor) Associated location IDs / a list of associated location IDs
* `bool` Whether this objective gets hidden if it's uncollectable

Whenever a map wants to update an objective or send a location, we update that into a call to `signalLocationId`, which should generally handle everything. If we keep handling counters separately, we will also update the counters with another function then.

Reply:
I'll note there's another case I'm not sure your test cases cover: what happens when a location is received without reaching that point in the mission? ie someone can `/sendLocation` or `/send` in the console, or someone else in a multiworld can `!collect`. If we spread all the check counter handling out in all the map scripts, we can't easily make updates to handle interactions like this. By providing all necessary data to the mod and handling things there, we only have to make such updates in one place.

There's also some path-dependency going on. If I start a mission with challenge locations enabled, then get sent stuff / find locations, I expect objectives to get crossed off. But if I start the mission with them disabled / everything collected, then I'd expect the objective to not show (assuming enablement and collection state are signaled through the same locations array).